### ISSUE TITLE
Adding checksum for safetensors format

### ISSUE THREAD

**Issue by liuliu** [huggingface/safetensors#220]
As safetensors gain wider usage in the community, it would be great if safetensors have some built-in checksum mechanism.

Recent case in point: some of my users reported incorrect weights reported (resulting NaN) from safetensors downloaded in civitai.com. If they download PyTorch format (PickleTensors), they will notice the file is corrupted because app refuses to import (zip has some built-in checksum mechanism). However, safetensors will just happily import with corrupted weights.

To make it backward compatible, I would suggest do this with conventions, for example, we just add to the __metadata__ key with: `"sha256sum": "hexcode"` that computes sha256sum from buffer starts (0 offset) to the very end.

What do you think?

**Comment by Narsil** [comment]
Thanks a lot for the suggestion:

tl;dr:

- It could be added in the `__metadata__`, but it's existence cannot be forced.
- It cannot be enforced on load for speed reasons. On-demand usage only (something like `safetensors.check_file(filename)`.



The problem is that calculating a checksum, especially over several gigabytes file is quite slow and so doing on load will lose any gains achieved by zero-copy (and actually probably being just worse than pure pickle). Torch itself doesn't check the zip checksums (proof follows)

One protection from corrupted files this was added https://github.com/huggingface/safetensors/pull/206 which should cover unfinished files. (It's not released yet)

Now, that being said, we could definitely start saving something like a checksum within the file to confirm the integrity of the data buffer (we can't check the whole file, since the header would be included in the checksum calculation and would mean we need to calculate the checksum with the already existing checksum, which is impossible to the best of my knowledge, while a fun little challenge).

This could exist in `__metadata__~. Also sha256 is relatively slow to calculate and since wouldn't be used for security, I think there are more appropriate faster algorithms

The thing is, we would never check it by default for the aforementioned reasons (it's slow).
But it could be nice for users having doubts to be able to run a check on a local file.
if that's the case, I think looking at the source hash would be even better (like here https://huggingface.co/hf-internal-testing/tiny-random-GPT2LMHeadModel/blob/main/pytorch_model.bin) but it's not always possible

This would only cover sanity checks, be on-demand and not by default.
Also since files already exist without this checksum, the existence of it cannot be enforced either


## Pickle is NOT protected from bitflips

Jus to prove it:

`flip.py`  https://gist.github.com/Narsil/6dd3c14a0819f6b35c51c4c4ec085bf7
Let's get this file: https://huggingface.co/hf-internal-testing/tiny-random-GPT2LMHeadModel/blob/main/pytorch_model.bin

```
$ sha256sum pytorch_model.bin
3b571099e33d420a168aeab3919c0b601df87cea689061bfd9a85baabd0c7f75  pytorch_model.bin

$ python flip.py pytorch_model.bin

$ sha256sum pytorch_model.bin
46c252f944618d137e00a980cece2f640300735674fd5428f1cbc4af08a3643e  pytorch_model.bin

$ python -c "import torch; weights = torch.load('pytorch_model.bin')"
```

It's probably for same reasons we cannot enforce checksums on loads, it's just slow, so for model files it's prohibitive.


## Hash for security: 

Doubling down, that this would not be about security, only to protect from random hardware/network error.

Hashes for security are meant to be outside of a file, and communicated from a different mechanism.

The reason is that it's means to be protecting from man-in-the-middle, if the man-in-the-middle can intercept both messages, then the checksum is useless, since the attacker can modify it as well. So checksums included within the file are worthless for security purposes. 
Just wanted to mention that loud and clear.

**Comment by liuliu** [comment]
Definitely, this is only for integrity check. And by build this inside this repo, (and generate the value by default, but not checking by default). It will help to populate this field in many shared safetensors files.

I mention SHA256 only for its ubiquitous availability on many platform and feel 32bit crc32 probably a bit ill-suited to cover this large data buffer although faster. We can investigate other checksum algorithms that is generally available. (also, while having some erasure coding inside the data buffer would prevent corruption at all, it seems will significantly increase the complexity of the implementation, that's why I didn't mention that as well).

**Comment by Narsil** [comment]
> (also, while having some erasure coding inside the data buffer would prevent corruption at all, it seems will significantly increase the complexity of the implementation, that's why I didn't mention that as well).|

Yes, I don't think it's worth it. Despite the files being large, EC needs to be done at another level for consumers requiring it IMO.
The size impact on the files is not negligible either I think. And depending on the criticity there are different levels of EC that could be done.

>  (and generate the value by default, but not checking by default). It will help to populate this field in many shared safetensors files.

Indeed. In general we need to be able to skip this because it's also slowing down writing, that's why we cannot enforce it.
Do you have any idea if CivitAI could share the sha256 too (this time for security purposes along with integrity ?)

**Comment by julien-c** [comment]
Personally i would tend to agree that checksumming should be done at another level

i.e. if you git clone a repo on huggingface.co git-lfs will checksum the sha256

**Comment by liuliu** [comment]
I tend to agree. The problem is the format itself have limited ways to ensure integrity naturally (i.e. it is mostly obvious in JPEG / PNG / ZIP if some bytes messed up (although this is a chance thing, as you proved before)) because other format tends to have structures while a blob of GiB payload have much less structures. So some "defense in depth" would help, even though we have out-of-band checks such as git-lfs, provided SHA256 by website operators, and ZFS checksum at disk level.

Anyway, while we are discussing, I think that I will implement some of these out-of-band checks as well as sanity check (isNaN) to minimize this kind errors in the future.

**Comment by Nerogar** [comment]
I already talked about this in the linked issue. But maybe it's better to discuss it here. What do you think about a new function that works similar to save(), but returns a stream instead of a full byte array? With that, anyone could add custom checksum calculations based on the actual saving logic.

Of course it would slow down the saving, because everything needs to be serialized twice. Once for the calculation. And once for the saving. But it opens up a lot more options.

It would also pe pretty efficient in terms of memory consumption, because only a small part of the data needs to be serialized at a time.

**Comment by Narsil** [comment]
Definitely an idea. 

I'm wondering how good we can make the API good (both give freedom and avoid super bad inefficiencies).

```python
for chunk in save_iter(tensors):
    # Hash + write to file ?
```

Would you be willing to open a PR for it ?

**Comment by liuliu** [comment]
> Maybe ?

Yeah, I would probably do it this way (return an iterator object), but probably keep the internal opaque? So in the future iterations, if the tensor is too large, you can break it down to return the iterator. It could literally just be
```
for chunk in save_chunks(tensors):
```
and the chunk is an object wrapped with a reference to the underlying Rust object rather than a direct copy of the underlying data so the lifetimes of such "chunk" is cheaper.

**Comment by Narsil** [comment]
@Nerogar use case would be to be able to `hash` the tensor contents into the metadata, so he would need access to the actual raw content of the tensors (to be able to hash).

The main thing, is that while you have all the RAM valid in `tensors` you don't have the order in which they would be written to know how to hash the contents, that's my proposal.
